sbt/src/test/scala/testpkg/ServerSpec.scala

```diff
@@ -216,13 +223,24 @@ case class TestServer(baseDirectory: File)(implicit ec: ExecutionContext) {
 
   @tailrec
   final def waitForString(num: Int)(f: String => Boolean): Boolean = {
-    if (num < 0) false
-    else
-      readFrame match {
-        case Some(x) if f(x) => true
-        case _ =>
-          waitForString(num - 1)(f)
+    if (num < 0) { throw new Exception("Retries are over.") } else {
```

Comment:
I think this could be simplified:
```scala
    if (num < 0) false
    else {
      val latch = new java.util.concurrent.CountDownLatch(1)
      Future(readFrame.filter(f).foreach(_ => latch.countDown()))
      latch.await(1, TimeUnit.SECONDS) || waitForString(num - 1)(f)
    }
```

Reply:
Alternatively, I'm not sure you even need a Future at all. Could this not be
```scala
  if (num < 0) false
  else readFrame.fold(false)(f) || { Thread.sleep(1000); waitForString(num - 1)(f) }
```
?

Reply:
while the two latter proposals basically change the semantics of this piece of code (i.e. they do not handle anymore the timeout properly). I do agree on moving to the `CountDownLatch` proposal, that is basically the desugared javaish original version :-)
One note is for keeping the exception in `if (num < 0)` because it's fundamental to properly terminate with timeout hanging tests.